pipelines/ui5-sap-cp/Jenkinsfile

```diff
@@ -21,15 +23,19 @@ node() {
 
   stage("Clone sources and setup environment"){
     deleteDir()
+    Map neoDeployConfiguration
     dir(APP_PATH) {
       checkout scm
       setupCommonPipelineEnvironment script: this, configFile: CONFIG_FILE
+      prepareDefaultValues script: this
+      neoDeployConfiguration = ConfigurationMerger.merge(ConfigurationLoader.stepConfiguration(this, 'neoDeploy'), ['host', 'account', 'credentialsId'],
+                                                         ConfigurationLoader.defaultStepConfiguration(this, 'neoDeploy'))
```

Comment:
Shouldn't the step configuration happen inside a given step, in this case inside neoDeploy? It's because of the validation stage later, correct? How about moving the validation of each tool into the relevant step as well, removing the validation stage from the pipeline?
The drawback would be that we don't detect a missing tool early that way... Personally though I would appreciate a simpler pipeline, moving this complexity to the steps.

Reply:
It is a benefit to have an early validation. Otherwise long running builds are launched which fails later at the end of build, but it is clear from the very beginning that the build cannot succeed since some prerequisites are missed. In case of neoDeploy eg we fails after the build step which might have been time consuming.

Checking prerequisites (again) inside the call method of a step might be beneficial also, since the circumstances might be changed from the beginning of the pipeline (hopefully not regarding some installed tools, but there might be other things we have to check).

All in all this is a difficult thing. 
* What happens if a step is switched of, maybe based on a branch pattern (eg for pull request branches some steps are skipped)? In this case the validation must also be skipped.
* What happens if parts of the builds are executed inside a docker container? In this case it does not make sense to check the environment on the machine running jenkins.

Maybe it would be beneficial to have a validate method as part of each step which can be called early in the pipeline and also at the beginning of the call method. With that validation and step execution in the narrower sense are closer together in one place in the code. Rather than having validation code in the pipeline and step execution code in the lib. Other option would be to have a validation twin for each step (eg neoDeploy and neoDeployValidate).

More comments and ideas from all participating developers would be appreciated.